[Worker support] test for no cors headers

madhur-tandon opened the pull request:
Closes https://github.com/pyscript/pyscript/issues/1312 and https://github.com/pyscript/pyscript/issues/1364

CC: @hoodmane and @antocuni

hoodmane commented on pyscriptjs/tests/integration/test_00_support.py:
```diff
@@ -32,6 +32,17 @@ def test_basic(self):
+        self.disable_cors_headers()
+        doc = """
+        <html>
+          <script>console.log(typeof SharedArrayBuffer === "undefined")</script>
```
I guess it's better just to use [crossOriginIsolated](https://developer.mozilla.org/en-US/docs/Web/API/crossOriginIsolated)

madhur-tandon replied:
done

antocuni commented on pyscriptjs/tests/integration/conftest.py:
```diff
@@ -130,6 +130,13 @@ class HTTPServer(SuperHTTPServer):
     """
 
+    def __init__(self, host_and_port, handler_class):
+        self.handler_class = handler_class
```
this is superfluous. `HTTPServer` is a subclass of `socketserver.SocketServer`, and the class is already stored as `RequestHandlerClass`:
https://github.com/antocuni/cpython/blob/deaf509e8fc6e0363bd6f26d52ad42f976ec42f2/Lib/socketserver.py#L201-L206

madhur-tandon replied:
thanks, done

antocuni commented on pyscriptjs/tests/integration/conftest.py:
```diff
@@ -130,6 +130,13 @@ class HTTPServer(SuperHTTPServer):
+        super().__init__(host_and_port, handler_class)
+
+    def disable_cors(self):
+        self.handler_class.use_cors = False
```
I would call this method `disable_cors_headers()`, it seems cleaner

madhur-tandon replied:
done

antocuni commented on pyscriptjs/tests/integration/conftest.py:
```diff
@@ -140,15 +147,18 @@ def run(self):
-def http_server(logger):
+def dev_server(logger):
     class MyHTTPRequestHandler(SimpleHTTPRequestHandler):
+        use_cors = True
```
maybe `enable_cors_headers` or `send_cors_headers`?

madhur-tandon replied:
used `enable_cors_headers `

antocuni commented on pyscriptjs/tests/integration/conftest.py:
```diff
@@ -162,7 +172,7 @@ def log_message(self, fmt, *args):
     thread.start()
 
-    yield base_url  # Transition to test here
+    yield server, base_url  # Transition to test here
```
I know I am the one who suggested it, but it was meant just as a quick hack to move forward, not as the final state.
With the current status we have:
  - a fixture called `dev_sever` which returns a tuple
  - an attribute `PyScriptTest.dev_server` which does *not* contain the fixture, but only the 2nd item

This is very confusing.

A better way to proceed is probably the following:
  - rename the `HTTPServer` class into `DevServer`
  - add an attribute to it called `addr` or `base_url`, which contains what is now in the `base_url` variable
  - then, the `dev_server` fixture can return an instance of `DevServer`, and `PyScriptTest` can just use `dev_server.base_url` to get the address

madhur-tandon replied:
Thanks, done!

antocuni commented on pyscriptjs/tests/integration/support.py:
```diff
@@ -126,6 +127,12 @@ def init_page(self, page):
+            self.router.use_cors = False
+        else:
+            self.dev_server.disable_cors()
+
```
see my comment above. If it's called `disable_cors_headers()` here, it should also be called `disable_cors_headers` in `dev_server`.
And probably in `router`, it should be called `use_cors_headers`.

madhur-tandon replied:
I am using `enable_cors_headers` to maintain consistency...
Or should I change it to `use_cors_headers` for `router`?

antocuni reviewed (changes requested):
The code in the PR is good, apart some naming/style minor issue which I highlighted in the comments.

But the big problem is the code which is NOT in the PR: what we want is that pyscript detects the "no cors" headers and display a nice error message, with a link to the documentation which explains what is the problem.
This is just the machinery to make it possible to write a test for that, but the machinery alone is useless.

hoodmane commented:
I don't think this does resolve #1312. The feature request is:

>  display a nice and error message (ideally with a link to a detailed page in our documentation) to explain what's the problem

When the headers are missing. As far as I can tell this patch only adjusts test code, it doesn't currently add any new error messages.

Also we should consider further improvements to synclink so that in same thread mode shared array buffer isn't needed.

hoodmane commented:
This is a useful step towards resolving #1312 though, it gets the test support code ready.

madhur-tandon commented:
Aah sorry, seems like I missed the part about introducing error messages in https://github.com/pyscript/pyscript/issues/1312

madhur-tandon commented on pyscriptjs/tests/integration/support.py:
```diff
@@ -216,7 +232,9 @@ def writefile(self, filename, content):
-        url = f"{self.http_server}/{path}"
+        url = f"{self.http_server_addr}/{path}"
+        if self.headers != self.REQUIRED_HEADERS:
+            self.page.evaluate("console.error('PyScript needs CORS headers to be set.')")
```
The error message can be changed OR perhaps there is a better way to do this.

antocuni commented:
@madhur-tandon I think there is a bit of confusion w.r.t the final goal. In theory, the current situation is the following:

1) PyScript uses `Synclink` which uses `SharedArrayBuffer`, which is available only under certain conditions
2) those conditions depends on the web server which servers the HTML page, and which we (PyScript OSS developers) have no control upon
3) if an user develops a pyscript app and deploys it to a webserver which does NOT send the required headers, the page would not work because it cannot find `SharedArrayBuffer`
4) *in that case*, we want to display a nice error message to the user, using an alert banner similar to what we display in case of `UserError`.
5) we want to write a test which checks that we actually display the error in case (4)

*HOWEVER*

I realized that we cannot write such a test currently, because situation (4) never arises: i.e., currently pyscript works perfectly fine even if we don't have CORS headers set. I *think* we will need it as soon as we merge PR #1333 , but not today.

@hoodmane I am very confused about what's happening on `main` currently. `Synclink` seems to have some fallback to use `ArrayBuffer` if `SharedArrayBuffer` is not available:
https://github.com/pyodide/synclink/blob/f26bfa33139bf9209cad9df5c03a4994fbc62c95/src/shared_array_buffer.ts#L1-L7

But then my question is: why did you have to modify `HTTPServer` in the synclink branch? In particular:
https://github.com/pyscript/pyscript/blob/c05195c0450cf0fa883cd208c39290780539e5df/pyscriptjs/tests/integration/conftest.py#L144-L156

If I comment out `self.send_my_header()` and revert `host` to `127.0.0.1`, the tests seems to work flawlessly. So I suppose that that change was not strictly necessary?

madhur-tandon commented:
@antocuni Although I have added the error message, it is not in the form of a `UserError` / alert banner.
If you have time, can you review this again?
Although PyScript works fine even if we don't have headers set, but I made sure to still check if our dev web server / fake server supplies them or not and we then do a `console.error` in case they do not.

Tagging @hoodmane too for a re-review as well.

The right entry point to check for these headers is the `goto()` function IMO
The commit where I added the error message is here: https://github.com/pyscript/pyscript/pull/1374/commits/97bc5d55acdcfb4f2153a7a15bdccdb7642e19e5

hoodmane commented:
> So I suppose that that change was not strictly necessary?

Sounds like.

antocuni commented:
> @antocuni Although I have added the error message, it is not in the form of a `UserError` / alert banner.

@madhur-tandon see my previous message:
https://github.com/pyscript/pyscript/pull/1374#issuecomment-1502885214

The source of the confusion is a big misunderstanding: the synclink PR added the CORS headers to our dev server, so I assumed they were necessary for running pyscript: and in that case, the point of #1312 was to make sure that we display a nice error message to the user.

But it turns out that it's not the case: as of now, the CORS headers are not needed and pyscript works perfectly fine without them. So the "proper" fix would be to revert the change and delete the extra code from the repo, not to add a test for an unneeded feature.

However, the situation is going to change soon: CORS headers *are* needed for web workers, so as soon as #1333 is merged, this PR becomes relevant again. Let's put it on hold until then.

madhur-tandon commented:
The tests that fail are kinda unrelated 😅

FabioRosado reviewed (approved):
Looks good to me 😄

hoodmane commented on pyscriptjs/src/main.ts:
```diff
@@ -160,6 +160,15 @@ export class PyScriptApp {
         this.config = loadConfigFromElement(el);
+        if (this.config.execution_thread === 'worker' && crossOriginIsolated === false) {
+            throw new UserError(
+                ErrorCode.BAD_CONFIG,
+                `execution_thread is set to "worker" but the following CORS headers are missing: ${JSON.stringify({
+                    'Cross-Origin-Embedder-Policy': 'require-corp',
+                    'Cross-Origin-Opener-Policy': 'same-origin',
+                })}`,
```
How do we know this? There might be a different reason that `crossOriginIsolated` is `false`, like the site being served over http (not https) or a browser bug. In general error messages should only make assertions of facts that we **know** are true. The reason that we are throwing the error is because crossOriginIsolated is false. So this is what we should report. We can then follow that with a suggestion that we think most likely because we are missing one or both of these two headers.

The following is more in the direction I'd think:
```
When execution_thread is "worker", the site must be cross origin isolated 
but crossOriginIsolated is false. To be cross origin isolated, it must be served with the following headers:

The problem may be that one or both of these are missing.
[in followup PR add link to discussion in DOCs]
```

hoodmane replied:
Also, in Chrome we have the option to set 
```
Cross-Origin-Embedder-Policy: credentialless
```
which is quite exciting because it allows you to load static resources from places that don't have the cross origin resource policy set. But it disabled authenticated requests so that's a tradeoff.

hoodmane reviewed (approved):
Looks good to me.

hoodmane commented:
After we merge let's open an issue to add a place in the docs somewhere with more info and put a link to it in the error message.

antocuni commented on pyscriptjs/tests/integration/test_01_basic.py:
```diff
@@ -32,6 +32,31 @@ def test_execution_thread(self):
+            """
+            <py-config>
+                magic = "unicorn"
+            </py-config>
```
why is this `<py-config>` needed?

madhur-tandon replied:
@antocuni I have removed it!

antocuni reviewed (approved):
LGTM
